scipy/optimize/cython_optimize/__init__.py

```diff
@@ -0,0 +1,630 @@
+
+     Similiar to :py:func:`~scipy.optimize.brentq` with a callback, ``f``, that
+     takes a double and a Python tuple ::
+
+        double brentq(callback_type_tuple f, double xa, double xb, tuple args,
+                      double xtol, double rtol, int iter, scipy_zeros_parameters *full_output)
+
+     :param callback_type_tuple f: the callback function
+     :param double xa: lower bound
+     :param double xb: upper bound
+     :param tuple args: extra parameters
+     :param double xtol: absolute tolerance
+     :param double rtol: relative tolerance
+     :param int iter: maximum number of iterations
+     :param scipy_zeros_parameters* full_output: optional output ``struct``, use ``NULL`` if not desired
+     :return: double
+
+Examples
+--------
+Usage of :py:mod:`~scipy.optimize.cython_optimize` requires the use of Cython
+to write callbacks that are compiled into native C. See the Cython Optimize
+Zeros API Jupyter notebook in the ``cython_optimize/Examples`` folder for usage
+examples. There are also several examples that solve the single-diode model of
+a solar cell in the ``cython_optimize/Examples`` folder. These examples are
+also used to test the different combinations of callbacks and root-finders.
+
+These are the basic steps to use :py:mod:`~scipy.optimize.cython_optimize`:
+
+1. Select a callback signature, for example:
+   :py:mod:`~scipy.optimize.cython_optimize.zeros_struct`
+2. Select the root finder, for example:
+   :py:func:`~scipy.optimize.cython_optimize.zeros_struct.newton`
+3. Create a Cython ``.pyx`` file that imports the zeros module with the
+   selected callback signature, create the callback(s) using the selected
+   signature, and call the selected root-finder passing the callback(s), any
+   extra arguments, and the other solver parameters ::
+
+       import math
+       cimport scipy.optimize.cython_optimize as coz
+
+       ARGS = {'C0': 1.0, 'C1': 0.7}
+       XTOL, RTOL, MITR = 1e-3, 1e-3, 10
+
+       ctypedef struct test_params:
+           double C0
+           double C1
+
+
+       cdef double f(double x, void *args):
+           cdef test_params *myargs = <test_params *> args
+           return myargs.C0 - math.exp(-(x - myargs.C1))
+
+
+       cdef double brentq_wrapper_example(dict args, double xa, double xb,
+                                          double xtol, double rtol, int mitr):
+           cdef test_params myargs = args  # Cython automatically casts dictionary to struct
+           return coz.zeros_struct.brentq(f, xa, xb, <test_params *> &myargs, xtol, rtol, mitr, NULL)
+
+
+       def brentq_example(args=ARGS, xa=0.5, xb=1.0, xtol=XTOL, rtol=RTOL, mitr=MITR):
+           return brentq_wrapper_example(args, xa, xb, xtol, rtol, mitr)
+
+
+       x = brentq_example()
+       # 0.6999942848231314
+
+4. If you want to call your function from Python, create a wrapper
+5. Create a Cython ``.pxd`` file if you need to export any Cython functions
+
+Full Output
```

Comment:
`full_output` is usually a design mistake in older Python APIs, I wouldn't copy it - better just include everything always.

Reply:
Actually, the `full_output` _is_ always available, the user just has to provide a structure for it.